sklearn/datasets/_openml.py

```diff
@@ -245,12 +246,8 @@ def _convert_arff_data(arff, col_slice_x, col_slice_y, shape=None):
     """
     arff_data = arff['data']
     if isinstance(arff_data, Generator):
-        if shape[0] == -1:
-            count = -1
-        else:
-            count = shape[0] * shape[1]
         data = np.fromiter(itertools.chain.from_iterable(arff_data),
-                           dtype='float64', count=count)
```

Comment:
Would like to highlight the reason for this change. If np fromiter is given an exact count, it will call `next(iterator)` only count times and hence, StopIteration is never raised on the stream. The only way to check md5 checksum after stream iteration is to determine end-of-stream via StopIteration paradigm [here](https://github.com/scikit-learn/scikit-learn/pull/14800/files#diff-93d3400ad9806682a64d5a6cc1c75b4fR465-R467) and then verify if checksums match.

I made this change so that it does the default behavior (which is consuming entire stream). The only catch here could be the edge case where arff may have more lines after the data section at some point, but a quick look at the standard makes me think that is cannot be a valid case.

Reply:
There are edge cases here by not setting count, which depends on the arrf format. Is it valid for count to have smaller than the data section?

Here is an alternative: `_load_arff_response`, we can consume the rest of the stream before passing back `parse_arff`:

```py
def _load_arff_response(...):
	...
	stream = _stream_checksum_generator(response)
	out = parse_arff(arff)

	# consume stream for validation
	for _ in stream:
		pass

	# do validation here
	if actual_md5_checksum.hexdigest() != md5_checksum:
		raise ValueError(...)

	return out
```